### PR TITLE
Implement modular sum constraint

### LINKED ISSUE #95: Implement affine constraints modulo m
**Remark 1**:This is for the constraints v2 after we finish #55.

**Remark 2**: the CoTenN paper/software does not implement this kind of constraint.

Using our current infrastructure, it would be simple to implement constraints of the form

    sum(w_i x_i) == k mod m

Although these are not the most common constraints in operations research, they are appear a lot in physics, e.g., parity conservation in the Ising model. 

This constraint becomes a DFA with `m` states. A sketch of an implementation would be somewhat like

```julia
struct SumModConstraint{T<:Real} <: AbstractConstraint
  weights::Dict{Int,T}
  mod::T
  rhs::T
end

function constraint_to_dfa(constraint::SumModConstraint{S}, nsites::Integer) where {S}
  (; weights, rhs) = constraint

  beyond    = rhs + one(S)
  states    = zero(S):constraint.mod
  alphabet  = S[0, 1]
  initial   = zero(S)
  accepting = Set([constraint.mod)

  id_dict = Dict((q, a) => q for q in states for a in alphabet)
  transitions = fill(id_dict, nsites)

  for site in constraint_sites(constraint)
    weight = weights[site]

    transitions[site] = Dict(
      (q, a) => (q + weight * a) % constraint.mod
      for q in states, a in alphabet
    )
  end

  return DFA(states, alphabet, initial, accepting, transitions)
end
```

### PR DESCRIPTION
**Pull request by iagoleal**
`SumModConstraint` with feasibility set satisfying

    sum(w[i] * x[i] for i in sites) == k mod m

This constraint is a new addition that is explicitly absent from the CoTenN paper.

Closes #95.

### CODE REVIEW

File: src/constraints.jl [hunk 1]
```diff
@@ -72,6 +72,38 @@ function SumConstraint(sites, weights, rhs; relation)
...
+`weights` must be the same length as `sites`.
+"""
+struct SumModConstraint{T<:Real} <: AbstractConstraint
+  weights::Dict{Int,T}
+  rhs::T
+  mod::T
+
+  function SumModConstraint{T}(sites, weights, rhs; mod) where {T<:Real}
+    site_vec   = validate_sites(sites)
+    weight_vec = validate_weights(map(w -> w % mod, weights))
```
Comment (bernalde) [inline src/constraints.jl]: Blocking: validate the modulus before arithmetic and use Euclidean modulo consistently. This currently applies `%` before checking `mod`, then feeds the residues through `SumConstraint`'s deliberately nonnegative validator. On this head, `SumModConstraint([1], [-1], -1; mod=3)` throws instead of normalizing to residue 2; `is_feasible([-1], SumModConstraint([1], [1], 2; mod=3))` returns false even though `-1 ≡ 2 (mod 3)`; lowering that case over `-1:1` produces an unknown target state `-1`; `mod=-3` and `mod=2.5` construct successfully; and `mod=0` leaks a `DivideError`. That contradicts #95's signed-normalization and `m >= 1` integer requirements. Validate `mod` first, validate the raw weights/rhs as integer-valued, normalize with `Base.mod`, use `Base.mod` in `is_feasible` and the DFA transitions, reject noninteger alphabets explicitly, and add regressions for signed weights/rhs, negative domains, and zero/negative/fractional moduli.

File: src/projection_mpo.jl [hunk 2]
```diff
@@ -232,6 +232,8 @@ Constraint site numbers use the same 1-based register indexing as `sites`.
 
 - [`SumConstraint`](@ref) uses a specialized exact integer partial-sum automaton.
 For a constraint with rhs `k`, its maximum bond dimension is `k+2`.
+- [`SumModConstraint`](@ref) uses a modular partial-sum automaton.
+  Its maximum bond dimension is is the remainder modulo `m`.
```
Comment (bernalde) [inline src/projection_mpo.jl]: Nonblocking: the dimension is `m`, not “the remainder modulo `m`,” and this also has a duplicated “is.” Please say that the modular DFA has `m` residue states and therefore bond dimension `m`. The focused projection test should assert `maxlinkdim(H) == constraint.mod`; its current `<= 3` check would still pass an unintended smaller construction.

**Review by bernalde — CHANGES_REQUESTED** [review]
## Review summary

Blocking issues: 1 inline finding. The modular arithmetic contract does not currently hold for signed integer inputs, and invalid moduli are not validated as required by #95.

Nonblocking issues: 1 inline documentation/test-contract correction.

Questions: none.

Verification on the exact PR head (`0c8012e`):

- `julia +1.10 --project=. -e 'using Test, TenSolver; include("test/constraints.jl")'`: 71/71 passed.
- `julia +1.10 --project=. -e 'using Test, Random, LinearAlgebra, TenSolver; include("test/utils.jl"); include("test/projection_mpo.jl")'`: 2001/2001 passed.
- A focused edge-case probe reproduced signed-weight rejection, incorrect negative-domain feasibility, an invalid negative DFA transition, acceptance of negative/fractional moduli, and a low-level `DivideError` for modulus zero.
- The same two targeted suites passed 71/71 and 2001/2001 on a clean merge result with current `main`; the edge-case probe reproduced the same defects there.
- GitHub reports all 10 current-head Actions checks successful (nine test matrix jobs and documentation), plus a successful `documenter/deploy` status.

Linked issue: the PR provides the requested specialized `m`-state DFA, but it misses #95's explicit positive-modulus validation and signed normalization requirements.

Coordination: open PR #120 touches all seven files in this PR, while #116 and #124 overlap projection/docs tests. The current merge with `main` is clean, but the eventual merge order should be rechecked.

Merge-readiness: not ready until the blocking modular-normalization and validation finding is addressed. `main` is currently unprotected, so GitHub does not enforce an approving-review gate.

I would not merge this until the blocking issues above are addressed.

**Comment by bernalde** [comment]
Addressed both review threads in [ba6086a](https://github.com/SECQUOIA/TenSolver.jl/commit/ba6086ab3136f17cbf75536bb59235463bb6a629).

Changes:

- Validate an integer-valued positive modulus before modular arithmetic.
- Validate raw weights/rhs as integer-valued, then normalize signed values with `Base.mod`.
- Use Euclidean modulo consistently in feasibility checks and DFA transitions; explicitly reject fractional domains.
- Correct the bond-dimension documentation and assert the exact `m`-state bond dimension.
- Add signed-weight/rhs, negative-domain, invalid-modulus, fractional-domain, and exact-bond regressions.

Verification:

- The new regressions failed on the reviewed implementation before the source fix.
- Constraint target: 85/85 passed.
- Projection-MPO target: 2067/2067 passed.
- Full `Pkg.test()`: passed (the suite retains its three pre-existing `Broken` real-model cases).
- Documentation build: passed.
- Clean merge result with current `main`: 85/85 constraint and 2067/2067 projection tests passed.
- Mutation checks: putting signed remainder back made the negative-feasibility test fail (84/85); putting remainder transitions back made the negative-domain projection test error on target state `-1`.

Comments intentionally not addressed: none.

CI:

- Documentation passed.
- The test matrix's project tests passed on every runner. The first attempt failed only when Aqua's persistent-task precompile check did not create `done.log` on macOS/Julia 1.
- One bounded rerun made that macOS job pass. The same Aqua `done.log` check then failed on Windows/Julia 1 after all project tests passed; fail-fast interrupted Windows/pre after its project tests and 10/10 completed Aqua checks passed. The other seven matrix jobs passed. This is classified as an infrastructure/precompile flake, and the workflow was not rerun a second time.

Remaining state: GitHub still reports `CHANGES_REQUESTED` until the submitted review is updated or dismissed; the two threads remain unresolved because this workflow does not resolve reviewer conversations. Open PRs #120, #116, and #124 still overlap parts of this change, so merge order should be rechecked.

File: src/constraints.jl (hunk 1 above)
Comment (bernalde) [reply src/constraints.jl]: Addressed in [ba6086a](https://github.com/SECQUOIA/TenSolver.jl/commit/ba6086ab3136f17cbf75536bb59235463bb6a629). The constructor now validates a positive integer-valued modulus before arithmetic, validates raw integer-valued weights/rhs, and normalizes with `Base.mod`; feasibility and DFA transitions use the same Euclidean convention, with fractional domains rejected explicitly. Regressions cover every reproduced edge case. I also put the signed-remainder behavior back temporarily; the negative-feasibility and negative-domain projection tests then failed, confirming they guard the fixes.

File: src/projection_mpo.jl (hunk 2 above)
Comment (bernalde) [reply src/projection_mpo.jl]: Addressed in [ba6086a](https://github.com/SECQUOIA/TenSolver.jl/commit/ba6086ab3136f17cbf75536bb59235463bb6a629). The documentation now states that the DFA has `m` residue states and bond dimension `m`, and the focused projection test asserts `maxlinkdim(H) == constraint.mod`.
